filament/backend/src/vulkan/VulkanPipelineCache.cpp

```diff
@@ -154,109 +142,99 @@ void VulkanPipelineCache::setDevice(VkDevice device, VmaAllocator allocator) {
 }
 
 bool VulkanPipelineCache::bindDescriptors(VkCommandBuffer cmdbuffer) noexcept {
-    VkDescriptorSet descriptors[VulkanPipelineCache::DESCRIPTOR_TYPE_COUNT];
-    bool bind = false, overflow = false;
-    getOrCreateDescriptors(descriptors, &bind, &overflow);
+    DescriptorMap::iterator descriptorIter = mDescriptorSets.find(mDescriptorRequirements);
 
-    if (overflow) {
-        return false;
+    // Check if the required descriptors are already bound.
+    if (DescEqual equals; UTILS_LIKELY(equals(mBoundDescriptor, mDescriptorRequirements))) {
+        assert_invariant(descriptorIter != mDescriptorSets.end());
+        descriptorIter.value().last_used = mCurrentTime;
+        return true;
     }
-    if (bind) {
-        LayoutBundle* layout = getOrCreatePipelineLayout();
-        vkCmdBindDescriptorSets(cmdbuffer, VK_PIPELINE_BIND_POINT_GRAPHICS,
-                layout->pipelineLayout, 0, VulkanPipelineCache::DESCRIPTOR_TYPE_COUNT,
-                descriptors, 0, nullptr);
+
+    // If a cached object exists, re-use it, otherwise a create a new one.
+    DescriptorCacheEntry* cacheEntry = UTILS_LIKELY(descriptorIter != mDescriptorSets.end()) ?
+            &descriptorIter.value() : createDescriptorSets();
+
+    // If a descriptor set overflow occurred, allow higher levels to handle it gracefully.
+    if (UTILS_UNLIKELY(cacheEntry == nullptr)) {
+        return false;
     }
+
+    cacheEntry->last_used = mCurrentTime;
+    mBoundDescriptor = mDescriptorRequirements;
+
+    vkCmdBindDescriptorSets(cmdbuffer, VK_PIPELINE_BIND_POINT_GRAPHICS,
+            getOrCreatePipelineLayout()->handle, 0, VulkanPipelineCache::DESCRIPTOR_TYPE_COUNT,
+            cacheEntry->handles.data(), 0, nullptr);
+
     return true;
 }
 
 void VulkanPipelineCache::bindPipeline(VkCommandBuffer cmdbuffer) noexcept {
-    VkPipeline pipeline;
-    if (getOrCreatePipeline(&pipeline)) {
-        vkCmdBindPipeline(cmdbuffer, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline);
+    PipelineMap::iterator pipelineIter = mPipelines.find(mPipelineRequirements);
+
+    // Check if the required pipeline is already bound.
+    if (PipelineEqual equals; UTILS_LIKELY(equals(mBoundPipeline, mPipelineRequirements))) {
+        assert_invariant(pipelineIter != mPipelines.end());
+        pipelineIter.value().last_used = mCurrentTime;
+        return;
     }
+
+    // If a cached object exists, re-use it, otherwise a create a new one.
+    PipelineCacheEntry* cacheEntry = UTILS_LIKELY(pipelineIter != mPipelines.end()) ?
+            &pipelineIter.value() : createPipeline();
+
```

Comment:
Do we need to check if `cacheEntry` is `nullptr` here (like you did in `bindDescriptors`)? Likewise maybe `bindPipeline` should also return a boolean.

Reply:
`createPipeline` must not be failed as I think.

Reply:
`createPipeline` returns `nullptr` if `vkCreateGraphicsPipelines` fails, though I suppose we also have an assert for this case.